### DISCUSSION TITLE
how to validate dynamically generated form with zod schema

### DISCUSSION THREAD

**Discussion by rayanazzam1991** [colinhacks/zod#1746] [Q&A] [answered]
i have multiple inputs generated after i got response from the server and i have to validate those inputs fields

**Comment by JacobWeisenburger** [comment]
Would you be able to give some example code of what you are trying to do?

**Reply by rayanazzam1991** [reply]
i am trying to build object shape dynamically like this 

```ts
let servicesValidationSchemaObj = ref({})
for (let i = 0; i < servicesChecked.value.length; i++) {
      let key = `service_price_${servicesChecked.value[i].service.id}`
      servicesValidationSchemaObj = {
        [key]: zod
          .preprocess(
            (input) => {
              const processed = zod.string({}).regex(/\d+/).transform(Number).safeParse(input);
              return processed.success ? processed.data : input;
            },
            zod
              .number({invalid_type_error: "Please enter a valid number"})
          )
      }

    }
    servicesValidationSchema.value = toFormValidator(zod
      .object(servicesValidationSchemaObj));
```
but i think i have problem with this, the servicesValidationSchemaObj  always have the latest value

**Reply by JacobWeisenburger** [reply]
it seems that you are overwriting the value of `servicesValidationSchemaObj` every iteration through the loop. I'm not really sure what this has to do with Zod. Could you explain more about the question you have about Zod?

**Reply by JacobWeisenburger** [reply]
This might be of some help too:
https://github.com/colinhacks/zod#coercion-for-primitives

**Reply by rayanazzam1991** [reply]
finally after i have spent couple of hours trying to make a dynamic schema that is generated in this for loop, but as you can see i can't make the servicesValidationSchemaObj dynamic because it's overriden in every iteration, so i solved the problem with ZOD dynamic schema and i now i am facing the problem of overwriting. if you can help with it please

**Reply by JacobWeisenburger** [reply]
Perhaps these links can help:
https://github.com/colinhacks/zod#records
https://github.com/colinhacks/zod#coercion-for-primitives

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const servicesValidationSchemaObj = z.record(
    z.string().startsWith( 'service_price_' ),
    z.coerce.number( { invalid_type_error: 'Please enter a valid number' } )
)

console.log(
    servicesValidationSchemaObj.safeParse( {
        service_price_1: '42',
        service_price_foo: 42
    } ).success // true
)
console.log(
    servicesValidationSchemaObj.safeParse( {
        1: '42',
        foo: 42
    } ).success // false
)
```

**Reply by rayanazzam1991** [reply]
really after you told me about record 
i tired this 
```ts
const validationSchemaRecord = zod.record(zod.string(),zod
      .preprocess(
        (input) => {
          const processed = zod.string({}).regex(/\d+/).transform(Number).safeParse(input);
          return processed.success ? processed.data : input;
        },
        zod
          .number({invalid_type_error: "Please enter a valid number"})
      ))
```
and it worked for me thx
